disktable.go

```diff
@@ -0,0 +1,650 @@
+	}
+
+	return ch
+}
+
+// exactMatch does an exact match for a key in the index. key must have been finalKey()'d.
+func (t *Table) exactMatch(key []byte, index *Index) lookupResult {
+	var result lookupResult
+
+	f := func(txn *badger.Txn) error {
+		item, err := txn.Get(key)
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(item.Key(), key) {
+			panic("known key search came up with different key: HUGE PROBLEM, HUGE!!!!!!")
+		}
+		value, err := item.ValueCopy(nil)
+		if err != nil {
+			return fmt.Errorf("problem getting value of key(%s): %w", string(key), err)
+		}
+		result = lookupResult{i: endian.Uint64(value)}
+		return nil
+	}
+
+	if err := index.db.View(f); err != nil {
+		result.err = err
+	}
+	return result
+
+}
+
+// multiMatch matches all keys in an index that allows duplicates. key must have been finalKey()'d.
+func (t *Table) multiMatch(key []byte, index *Index, ch chan lookupResult) {
+	defer close(ch)
+
+	keyData, err := getPrefixSuffix(key)
+	if err != nil {
+		ch <- lookupResult{err: err}
+		return
+	}
+
+	tempBuff := make([]byte, 8)
+	err = index.db.View(
+		func(txn *badger.Txn) error {
+			it := txn.NewIterator(badger.DefaultIteratorOptions)
+			defer it.Close()
+
+			for it.Seek(keyData.prefix); it.ValidForPrefix(keyData.prefix); it.Next() {
+				item := it.Item()
+
+				foundKeyData, err := getPrefixSuffix(item.Key())
+				if err != nil {
+					return fmt.Errorf("multiMatch of index(%s) had bad key(%v): %w", index.Name, item.Key(), err)
+				}
+				// It is possible to have a prefix match for something not exactly the same.
+				// But we are looking for exact prefix matches followed by a number followed by
+				// our key size. By matching prefix and key size, we can be sure of an exact match.
+				// If not, throw out the data.
+				if len(foundKeyData.prefix) != len(keyData.prefix) {
+					continue
+				}
+				b, err := item.ValueCopy(tempBuff)
+				if err != nil {
+					return err
+				}
+				ch <- lookupResult{i: endian.Uint64(b)}
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		ch <- lookupResult{err: err}
+	}
+}
+
+func (t *Table) FetchAll(ctx context.Context) (chan Result, error) {
+	stream := t.primary.db.NewStream()
+	stream.NumGo = 16
+	stream.LogPrefix = "FetchAll.Streaming"
+
+	ch := make(chan Result, 1)
+	stream.Send = func(buf *z.Buffer) error {
+		list, err := badger.BufferToKVList(buf)
+		if err != nil {
+			return err
+		}
+
+		for _, kv := range list.Kv {
+			// TODO(jdoak): This doesn't seem to matter.
+			v := make([]byte, len(kv.Value))
+			copy(v, kv.Value)
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case ch <- Result{Value: v}:
+			}
+		}
+		return nil
+	}
+
+	go func() {
+		defer close(ch)
+		if err := stream.Orchestrate(context.Background()); err != nil {
+			ch <- Result{Err: err}
+			return
+		}
+	}()
+
+	return ch, nil
+}
+
+// ByteSlice2String coverts bs to a string. It is no longer safe to use bs after this.
+// This prevents having to make a copy of bs.
```

Comment:
the second sentence is unclear to me. also, the major feature of this function appears to be doing a conversion w/o a copy. how about the following instead:

```suggestion
// ByteSlice2String coverts bs to a string without incurring the cost of a copy of the given
// []byte parameter.
// 
// You may pass a specific []byte value to this function at most once. 
// It's unsafe to call this function with the same []byte value two or more times.
```

Reply:
I've altered some.  But the part about passing it more than once isn't true.

You can pass that []byte value to this 100 times and be safe, because string is immutable.  No change to any of the strings will cause a problem in another string, because the []byte will be copied.  I have no idea why you would do this, but should be fine.

The only thing you can't do is modify the []byte that you passed in, that would be bad.

Reply:
ah, ok. I misunderstood the comment in the first place, then